aten/src/ATen/native/cuda/UnfoldBackwardKernel.cu

```diff
@@ -0,0 +1,231 @@
+      auto idx_last_dim = *reinterpret_cast<int64_t*>(idx_last_dim_ptr + offsets[3]);
+
+      auto grad_out_idx_dim = idx_dim * step + idx_last_dim;
+      grad_out_data[grad_out_idx_dim * grad_out_dim_stride] = *grad_in_data;
+    };
+
+    _launch_unfold_backward_kernel<num_threads, thread_work_size>(iter.numel(), loop);
+  }
+  else {
+    auto offset_calc = make_offset_calculator<3>(iter);
+
+    // The algorithm is: for each index in grad_out find
+    // the elements contributing to it and sum them up.
+    // Note: the algorithm does not require any synchronization.
+    auto loop = [=]C10_DEVICE(int i) {
+      auto offsets = offset_calc.get(i);
+
+      auto* __restrict__ grad_out_data = reinterpret_cast<scalar_t*>(grad_out_ptr + offsets[0]);
+      auto* __restrict__ grad_in_data = reinterpret_cast<scalar_t*>(grad_in_ptr + offsets[1]);
+
+      auto idx_dim = *reinterpret_cast<int64_t*>(idx_dim_ptr + offsets[2]);
+
+      // left_fold potentially intersecting with idx_dim
+      // is either (idx_dim - size) / step or the next integer.
+      int64_t left_fold_idx = (idx_dim > size) ? (idx_dim - size) / step : 0;
+      if (!(left_fold_idx * step <= idx_dim && idx_dim < left_fold_idx * step + size)) {
+        ++left_fold_idx;
+      }
+
+      auto right_fold_idx = idx_dim / step;
+      right_fold_idx = (right_fold_idx >= grad_in_dim_size)
+        ? (grad_in_dim_size - 1) : right_fold_idx;
+
+      for (auto fold_idx = left_fold_idx; fold_idx <= right_fold_idx; ++fold_idx) {
+        auto idx_last_dim = idx_dim - fold_idx * step;
+        *grad_out_data += grad_in_data[fold_idx * grad_in_dim_stride
+                                    + idx_last_dim * grad_in_last_dim_stride];
+      }
+
+    };
+
+    _launch_unfold_backward_kernel<num_threads, thread_work_size>(iter.numel(), loop);
+  }
+}
+
+void unfold_backward_cuda_kernel(
+  Tensor& grad_out,
+  const Tensor& grad_in,
+  int64_t dim,
+  int64_t size,
+  int64_t step
+) {
+  dim = maybe_wrap_dim(dim, grad_out.dim());
+  // last dim stores the folds
+  auto last_dim = maybe_wrap_dim(-1, grad_in.dim());
+
+  auto grad_in_dim_stride = ensure_nonempty_stride(grad_in, dim);
+  auto grad_in_last_dim_stride = ensure_nonempty_stride(grad_in, last_dim);
+  auto grad_in_dim_size = ensure_nonempty_size(grad_in, dim);
+
+  auto grad_out_dim_stride = ensure_nonempty_stride(grad_out, dim);
+
+  auto is_step_ge_size = (step >= size);
+
+  TensorIterator iter;
+  if (is_step_ge_size) {
+    iter = _make_unfold_backward_iter_over_grad_in(
+      grad_out, grad_in, dim, size, step
+    );
+  }
+  else {
+    iter = _make_unfold_backward_iter_over_grad_out(
+      grad_out, grad_in, dim, size, step
+    );
+  }
+
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(
+    at::ScalarType::Half, at::ScalarType::Bool, at::ScalarType::BFloat16,
+    iter.dtype(),
+    "unfold_backward_cuda", [&] {
+      _unfold_backward_internal_kernel<scalar_t>(
```

Comment:
std::complex is not supported on cuda, so we use thrust::complex instead. so I think you could do this:
```
using thrust_t = typename ztype_cuda<scalar_t>::thrust_t;
_unfold_backward_internal_kernel<thrust_t>(...)
``` similar to what's being done [here](https://github.com/pytorch/pytorch/blob/69e2f1aaff9614bd801b34b0c9b4cc8305ee9a61/aten/src/ATen/native/cuda/UnaryFractionKernels.cu)

cc. @zasdfgbnm

Reply:
Thanks, that worked!